박스오피스 [STEP 2] Dasan, Whales

DasanKim opened the pull request:
안녕하세요, 븨븨(@YebinKim)!! 💃🏻
목표보다 찌끔 늦어버린 Step2 PR입니다 🥺
낯선 개념에, 코드 구현하는 것도 어려워서 많이 고민하고
많이 헤매기도했습니다!!
부족한 부분 다양하게 짚어주시면 더 열심히 공부해보겠습니다!!

잘 부탁드립니다 🙇🏻‍♀️

---

## 🤹🏻 고민했던 점
### 🔹 서버에서 받아온 data decoding
🚨 **문제점** <br>

- 서버에서 받아온 JSON데이터를 swift에서 사용할 수 있는 타입으로 decoding을 해야하는데, 어느 부분에서 decoding하는 것이 좋을지 고민하였습니다.
- 처음엔 `NetworkManager` 타입 내 `fetchData`에서 data를 받아오자마자 **타입을 분별하여 `BoxOffice` 또는 `Movie` 타입으로 decoding** 해주었습니다.
- 나중에 `BoxOffice` 또는 `Movie` 타입 외에 **더 많은 타입들이 생겼을 때**, `fetchData` 메서드가 비대해질 것을 대비하여 위 기능을 분리하였습니다.
    ```swift 
    struct NetworkManager {    
        func fetchData(url: String, queryItems: [URLQueryItem]) {
            (...)

            let task = URLSession.shared.dataTask(with: url) { data, response, error in
                (...)
                guard let data = data else { return }

                if let boxOffice: BoxOffice = try? DecodingManager.decodeJSON(data: data) {
                    print (boxOffice)
                } else if let movie: Movie = try? DecodingManager.decodeJSON(data: data) {
                    print (movie)
                } else {
                    return
                }
            }
            task.resume()
        }
    }
    ```

💡 **해결방법** <br>
- `fetchData` 메서드에 클로저 `completionHandler` 매개변수를 추가하여 해당 data의 분별 및 decoding을 completionHandler에서 처리하도록 하였습니다.

    ```swift
    struct NetworkManager {   
        func fetchData(url: URL?, completionHandler: @escaping (Data) -> ()) {
            (...)

            let task = URLSession.shared.dataTask(with: url) { data, response, error in
                (...)

                guard let data = data else { return }

                completionHandler(data)                                            
            }
            task.resume()
        }
    }
    ```
    ```swift
    //ViewController.swift
    let networkManager = NetworkManager()

    networkManager.fetchData(url: KobisOpenAPI.boxOffice(targetDate: "20220102").url) { data in
        guard let decodedData = try? DecodingManager.decodeJSON(type: BoxOffice.self, data: data) else { return }
        print(decodedData)
    }
    ```
 
### 🔹 API 프로토콜, KobisOpenAPI 타입

🚨 **문제점** <br>
- API를 `BoxOffice`와 `Movie` 두 가지 타입으로 받아오면서 URL을 작성하는 부분의 코드가 중복으로 구현되어 좀 더 효율적이고 유연한 코드를 위해 고민했습니다.

💡 **해결방법** <br>
- `KobisOpenAPI`이라는 열거형 타입을 생성해서 `boxOffice`와 `movie` 케이스를 만들어놓고 `API 프로토콜`을 채택해 URL의 필수 구성요소를 구현할 수 있도록 코드를 작성했습니다.
 
    ```swift
    protocol API {
        var baseURL: String { get }
        var path: String { get }
        var query: [URLQueryItem] { get }
    }

    enum KobisOpenAPI {
        case boxOffice(targetDate: String)
        case movie(movieCode: String)
    }

    extension KobisOpenAPI: API {
        var baseURL: String {
            return "http://www.kobis.or.kr"
        }

        var path: String {
            switch self {
            case .boxOffice:
                return "/kobisopenapi/webservice/rest/boxoffice/searchDailyBoxOfficeList.json"
            case .movie:
                return "/kobisopenapi/webservice/rest/movie/searchMovieInfo.json"
            }
        }

        var query: [URLQueryItem] {
            switch self {
            case .boxOffice(let targetDate):
                return [URLQueryItem(name: "key", value: key),
                        URLQueryItem(name: "targetDt", value: targetDate)]
            case .movie(let movieCode):
                return [URLQueryItem(name: "key", value: key),
                        URLQueryItem(name: "movieCd", value: movieCode)]
            }
        }

        var key: String {
            return "fb92260b76d9959bbf3ab69c991d8985"
        }

        var url: URL? {
            return URL(baseURL, path, query)
        }
    }
    ```
</br>

- 또한 `URL 구조체`를 확장하여 현 프로젝트에서 보기 쉽도록 `initializer`를 구현해주었습니다.
    ```swift
    extension URL {
        init?(_ baseURL: String, _ path: String, _ query: [URLQueryItem]) {
            var urlComponents = URLComponents(string: baseURL)
            urlComponents?.path = path
            urlComponents?.queryItems = query

            guard let url = urlComponents?.url else {
                return nil
            }

            self = url
        }
    }
    ```

## 🎙️ 조언을 얻고 싶은 점
### 🔸 Network 관련 타입
- Network와 관련된 타입들을 처음 작성하다보니, 기능 분리나 타입 분리 등에 많이 고민이 있었습니다. 이번 프로젝트를 진행하면서 기능 분리를 해본다고 했지만 기준이 명확하지 않아 쉽지 않았는데요😂 혹 Network와 관련된 타입들을 작성할 때 참고할 만한 기준이 있을까요?

## 🔗 참고링크
- [🍎Apple Docs: URLSession](https://developer.apple.com/documentation/foundation/urlsession)
- [🍎Apple Docs: Fetching Website Data into Memory](https://developer.apple.com/documentation/foundation/url_loading_system/fetching_website_data_into_memory)
- <Img src = "https://github.com/WhalesJin/ios-bank-manager/assets/124643545/d1df2d8a-6667-438d-9643-aab8a83a4754" width="20"/> [Alamofire](https://github.com/Alamofire/Alamofire/blob/master/Source/Alamofire.swift)
- <Img src = "https://github.com/WhalesJin/ios-bank-manager/assets/124643545/d1df2d8a-6667-438d-9643-aab8a83a4754" width="20"/> [Moya](https://github.com/Moya/Moya/tree/master/Sources/Moya)

DasanKim committed:
✨feat: Movie, MovieInformationResult, MovieInformation 타입 생성

WhalesJin committed:
🩹chore: 각각의 관련 타입들 BoxOffice, Movie 파일로 병합

DasanKim committed:
✨feat: DecodingManager내 decodeJSON(data:) 매서드 구현

DasanKim committed:
✨feat: NetworkManager 타입 생성 및 구현

WhalesJin committed:
🐛fix: Movie 타입 CodingKey 오류 수정 및 네이밍 변경

WhalesJin committed:
♻️refactor: fetchData 메서드에 completionHandler 매개변수 추가

DasanKim committed:
✨feat: API 프로토콜 생성, KobisOpenAPI 열거형 타입 생성 및 구현

WhalesJin committed:
🩹chore: API 프로토콜 및 KobisOpenAPI 타입 파일 분리

WhalesJin committed:
✨feat: URL 구조체의 새로운 initializer 구현

DasanKim committed:
♻️refactor: decodeJSON(data:)메서드에 type 매개변수 추가

DasanKim committed:
🩹chore: KobisOpenAPI 타입 케이스 연관값의 매개변수 이름 추가

YebinKim commented on BoxOffice/Controller/ViewController.swift:
```diff
@@ -11,9 +11,21 @@ class ViewController: UIViewController {
 
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
```
p3;

쓸모없는 주석을 삭제하셨군요! 👏👏
뷰컨트롤러의 이름도 디폴트 이름을 사용하는 것이 아닌 해당 뷰컨트롤러의 역할에 맞게 수정해보면 어떨까요?

WhalesJin replied:
관련 커밋: e5e1f7c8544020b417ca58e35b9d71b9b1e51720

프로젝트에 맞게 `BoxOfficeViewController`로 수정하였습니다!

YebinKim commented on BoxOffice/Model/API.swift:
```diff
@@ -0,0 +1,14 @@
+protocol API {
+    var baseURL: String { get }
+    var path: String { get }
+    var query: [URLQueryItem] { get }
```
p2;

Collection 타입에는 복수형의 프로퍼티 이름이 필요해보여요 👀

DasanKim replied:
- 관련 커밋: ce6d2ee4138e385e832e6f59cd97920e87df2708

query의 복수형인 queries으로 수정하였습니다😆

YebinKim commented on BoxOffice/Model/KobisOpenAPI.swift:
```diff
@@ -0,0 +1,47 @@
+        case .boxOffice:
+            return "/kobisopenapi/webservice/rest/boxoffice/searchDailyBoxOfficeList.json"
+        case .movie:
+            return "/kobisopenapi/webservice/rest/movie/searchMovieInfo.json"
```
p4;

`path` 내에서 중복되는 경로(`/kobisopenapi/webservice/rest`)가 있는 것으로 보여요 👀
중복되는 경로를 줄여볼 수 있을까요?

WhalesJin replied:
관련 커밋: c2f2a9a8a252e7f8cc056259543e68d11bf346e0

공통된 `path`에 `switch` 문을 이용하여 서비스마다 상이한 `path`를 붙여주는 것으로 수정하였습니다.

YebinKim commented on BoxOffice/Model/KobisOpenAPI.swift:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+    
+    var query: [URLQueryItem] {
```
p3;

query 를 포함해 외부에 노출되지 않을 프로퍼티는 private 처리가 필요해보여요 👀

DasanKim replied:
- `프로토콜`의 프로퍼티나 메서드에는 `접근 제어`를 사용할 수 없다는 것을 알게되었습니다. 그리하여 접근 제어와 프로토콜  중 어떤 것을 적용할 것인가에 대해 고민해보았습니다. 
- 프로토콜은 범용적으로 사용(구조체, `열거형`에도 적용가능)할 수 있다는 점과, **추후 테스트에도 용이**하다는 장점을 가지고 있는데, 이것에 초점을 두어 프로토콜을 구현하는 방향으로 코드를 작성했습니다:)

YebinKim commented on BoxOffice/Model/KobisOpenAPI.swift:
```diff
@@ -0,0 +1,47 @@
+    
+    var key: String {
+        return "fb92260b76d9959bbf3ab69c991d8985"
+    }
```
질문;

키값이 외부에 노출되어 있네요! 이 상태라면 누군가 서버의 키 값을 탈취하는 위험이 있을 거 같은데 혹시 이 보안상의 위험에 대해 생각해보셨을까요?

WhalesJin replied:
`접근 제어`를 하거나, `namespace`로 빼는 정도까지만 생각했던 것 같습니다🥲
하지만 조금 더 깊이 생각해보니 외부에서 접근을 못한다 하더라도 github에서는 코드를 다 볼 수 있으므로 근본적인 해결방법이 아닌 것 같아 API Key를 숨기는 방법을 알아보았습니다. 여러 방법이 있었지만 해당 [블로그](https://velog.io/@leedool3003/iOS-API-Key-숨기기)를 참고하여 진행하였습니다.

#### Key 숨기기

1. 새로운 `plist`를 만들어 KOBIS_API_KEY라는 이름의 `Key`와 KOBIS API KEY VALUE의 `Value`를 생성후, push 해주었습니다.
2. 후에 **더이상 새로운 `plist`파일을 추적하지 못하도록** `.gitignore` 파일 내에 새로운 plist path를 추가해준 뒤 push 해주었습니다.
3. 더 이상 추적하지 못하도록 한 뒤, `KOBIS API KEY VALUE`로 적혀있던 Value에 `실제 API Key`를 넣어주었습니다.

#### Key 사용하기
- Main Bundle를 사용하여 API Key값을 가져오도록 하였습니다!

##### Bundle 이란?
- `실행 가능한 코드`와 그 코드가 사용하는 `자원`을 포함하고 있는 디렉토리
- 가지고 있는 내용: info.plist, assets, string 파일 등
- `mainBundle`은 **앱이 실행되는 코드가 있는 Bundle 디렉토리에 접근**할 수 있게 도와주는 bundle

YebinKim replied:
번들에 대한 지식까지!! 알찬 공유 감사합니다 😆

YebinKim commented on BoxOffice/Model/API.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+protocol API {
```
p2;

프로토콜을 활용해 API 타입들을 추상화하셨군요! 🤩👏

다만, 현재 이름이 Swift API Design Guidelines 를 따르지 않고 있는 것으로 보여서 프로토콜을 채택할 타입이 수행할 역할을 기준으로 더 적절한 이름을 작성해볼 수 있을까요?
Swift API Design Guidelines의 [Strive for Fluent Usage](https://www.swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage)에서도 프로토콜을 통해 특정 기능을 수행할 수 있음을 명시하고자 하는 경우 접미사로 `-able` 등을 활용할 것을 권장하고 있습니다.
> Protocols that describe a capability should be named using the suffixes able, ible, or ing (e.g. Equatable, ProgressReporting).

DasanKim replied:
- 관련 커밋: e5e1f7c8544020b417ca58e35b9d71b9b1e51720

`URL`을 구성하는 요소들을 가지고 있게 하는 프로토콜이라 `URLConfigurable`으로 네이밍 수정하였습니다:)

YebinKim commented on BoxOffice/Model/NetworkManager.swift:
```diff
@@ -0,0 +1,34 @@
+        
+        let task = URLSession.shared.dataTask(with: url) { data, response, error in
+            if let error = error {
+                print(error)
```
p4;

로그로서 남기고자 한 것일까요? 로그로 남기고자 한다면 더 구체적인 정보를 남기는 것이 필요해보여요. (파일의 위치, 호출 메서드 등)
또는, print 메서드 대신 로깅과 관련된 메서드를 활용해보는 것도 추천드립니다 😄

YebinKim reviewed (changes requested):
다산, 웨일 안녕하세요 😄
Step2 진행까지 고생 많으셨습니다~

서버에서 받아 온 두 가지 타입의 데이터를 처리하기 위해 많은 고민을 하셨군요!
덕분에 의도하셨던 바와 같이 중복 코드를 많이 줄일 수도 있었고, API와 통신할 모델을 서버와 통신할 객체가 모르게 함으로써 비즈니스 로직(사용자에게 제공하는 기능과 관련된 로직)과 서비스 로직(네트워킹 또는 DB를 담당하는 로직)이 잘 분리되기도 했다고 생각해요 👍

다만, `NetworkManager` 와 `DecodingManager`가 Model 에 해당하는지는 조금 의문이 들어요 🤔
**더 적절한 위치의 폴더로 이동시켜볼 수 있을까요?**

---

### 1. Network 관련 타입

참고할만한 기준은 제시하기 어렵지만.. 현재는 수행하고 있는 기능인 쿼리 생성과 네트워킹(반드시 성공 기준)으로는 최대한 분리되었다고 생각했어요!
여기서 좀 더 기능을 나눠보자면 `서버 상태 및 클라이언트 에러 처리`와 `응답값의 추상화` 가 추가되면 더 역할을 나눠볼 수 있을 거 같아요 😄

그리고 어떤 기준을 세우고 작업하셨을지도 궁금한데, 혹시 이 부분은 회의실에서도 얘기 나눠볼 수 있을까요?

---

이번 스텝의 핵심 경험은 아래와 같았는데요.
<img width="305" alt="step2" src="https://github.com/yagom-academy/ios-box-office/assets/38066039/2c69053b-381d-45cf-b75a-93ed30489d4e">

네이밍과 관련해 p2 코멘트가 있어서 Request Changes 로 드렸습니다~
혹시 코멘트 내용 및 수정 방향에 대한 추가 설명이 필요하다면 언제든 질문주세요 😄🙌

WhalesJin committed:
♻️refactor: 쿼리 네이밍 및 key 접근제어 수정

WhalesJin committed:
♻️refactor: path 중복 로직 수정

DasanKim committed:
🩹chore: Utility 파일 그룹 생성 및 DecodingManager, NetworkManger 파일 이동

DasanKim committed:
🩹chore: viewController 및 API 프로토콜 네이밍 수정

WhalesJin committed:
✨feat: Error 그룹화

- Error 폴더 생성
- LocalizedError 프로토콜 준수하도록 DataError 수정
- LocalizedError 프로토콜 채택한 NetworkError 생성 및 구현

DasanKim committed:
♻️refactor: 네트워크 관련 에러 처리 수정

DasanKim committed:
🩹chore: 네이밍 수정 및 URL+ 파일 그룹 이동

DasanKim committed:
🩹chore: api key를 위한 .gitignore 수정

DasanKim committed:
✨feat: KOBIS_API_KEY.plist 파일 생성

DasanKim committed:
🩹chore: api key를 위한 .gitignore 재수정

DasanKim committed:
🩹chore: .gitignore 리셋

DasanKim committed:
🩹chore: .gitignore내 KobisAPIKey.plist path 수정

DasanKim committed:
🩹chore: Resource 폴더 생성

DasanKim committed:
🩹chore: 기존 KobisAPIKey 폴더 이동

DasanKim committed:
🩹chore: KobisAPIKey.plist Resource 그룹에 재생성

DasanKim committed:
🩹chore: .gitignore 내 KobisAPIKey path 리셋

DasanKim committed:
🩹chore: .gitignore에 KobisAPIKey path 수정

WhalesJin committed:
🐛fix: Info.plist의 path 수정

WhalesJin committed:
✨feat: Bundle extension으로 KOBIS_API_KEY 프로퍼티 생성

YebinKim commented on BoxOffice/Error/DataError.swift:
```diff
@@ -0,0 +1,22 @@
+    case failedDecoding
+    
+    var errorDescription: String {
+        switch self{
```
```suggestion
        switch self {
```

띄어쓰기가 하나 누락되어 보여요 👀

WhalesJin replied:
오 놓쳐버렸네요🫢!!
알려주셔서 감사합니다 수정하였습니다!!!

관련 커밋 : dc3a972182dec3fac2bdaa0a721354450a29f7ab

YebinKim commented on BoxOffice/Controller/BoxOfficeViewController.swift:
```diff
@@ -0,0 +1,54 @@
+                } catch DataError.notFoundAsset {
+                    print(DataError.notFoundAsset.localizedDescription)
+                } catch DataError.failedDecoding {
+                    print(DataError.failedDecoding.localizedDescription)
+                } catch {
+                    os_log("알 수 없는 오류입니다.", type: .default)
```
p4;

로그를 남기는 용도의 메서드가 두 개인 것으로 보여요 👀
특별히 구분하고자 하는 의도가 없다면 하나로 통일시키는 게 어떨까요?

YebinKim reviewed (approved):
다산, 웨일 안녕하세요~
Step2 수정까지 고생 많으셨습니다 😄

회의실에서 얘기 나눈 대로 이번 스텝에서 경험하고자 하는 내용은 모두 충분히 달성되었다고 생각해서 이대로 머지 진행할게요!
다음 스텝 진행도 화이팅입니다 🚀🎉
